prov/verbs/src/fi_verbs.c

```diff
@@ -557,110 +590,192 @@ static int fi_ibv_rai_to_fi(struct rdma_addrinfo *rai, struct fi_info *fi)
  	return 0;
 }
 
-static int fi_ibv_fill_info_attr(struct ibv_context *ctx, struct ibv_qp *qp,
-				 const struct fi_info *hints,
-				 struct fi_info *fi)
+static int fi_ibv_get_device_attrs(struct ibv_context *ctx,
+		const struct fi_info *hints, struct fi_info *info)
 {
-	struct ibv_qp_init_attr qp_init_attr;
+	struct ibv_pd *pd;
+	struct ibv_cq *cq;
+	struct ibv_qp *qp;
+	struct ibv_qp_init_attr init_attr;
 	struct ibv_device_attr device_attr;
 	struct ibv_port_attr port_attr;
-	struct ibv_qp_attr qp_attr;
-	union ibv_gid gid;
-	size_t name_len;
-	int ret;
-
-	*(fi->fabric_attr) = verbs_fabric_attr;
-	*(fi->domain_attr) = verbs_domain_attr;
-	*(fi->ep_attr)	   = verbs_ep_attr;
-	*(fi->tx_attr)	   = verbs_tx_attr;
-	*(fi->rx_attr)	   = verbs_rx_attr;
-	
-	if (!(fi->fabric_attr->prov_name = strdup(VERBS_PROV_NAME)))
-		return -FI_ENOMEM;
+	int ret = 0;
 
-	if (!ctx) {
-		if (!(fi->fabric_attr->name = strdup(VERBS_ANY_FABRIC)))
-			return -FI_ENOMEM;
+	pd = ibv_alloc_pd(ctx);
+	if (!pd)
+		return -errno;
 
-		return 0;
+	cq = ibv_create_cq(ctx, 1, NULL, NULL, 0);
+	if (!cq) {
+		ret = -errno;
+		goto err1;
 	}
 
-	if (qp) {
-		ret = ibv_query_qp(qp, &qp_attr, IBV_QP_CAP, &qp_init_attr);
-		if (ret)
-			return -ret;
-		fi->tx_attr->inject_size = qp_attr.cap.max_inline_data;
+	memset(&init_attr, 0, sizeof init_attr);
+	init_attr.send_cq = cq;
+	init_attr.recv_cq = cq;
+	init_attr.cap.max_send_wr = 1;
+	init_attr.cap.max_recv_wr = 1;
+	init_attr.cap.max_send_sge = 1;
+	init_attr.cap.max_recv_sge = 1;
+
+	if (hints && hints->tx_attr && hints->tx_attr->inject_size) {
+		init_attr.cap.max_inline_data = hints->tx_attr->inject_size;
 	} else {
 		fi_read_file(FI_CONF_DIR, "def_inline_data",
-			def_inline_data, sizeof def_inline_data);
-		fi->tx_attr->inject_size = atoi(def_inline_data);
+		def_inline_data, sizeof def_inline_data);
+		init_attr.cap.max_inline_data = atoi(def_inline_data);
+	}
+
+	init_attr.qp_type = IBV_QPT_RC;
+
+	qp = ibv_create_qp(pd, &init_attr);
+	if (!qp) {
+		ret = -errno;
+		goto err2;
 	}
 
-	ibv_query_gid(ctx, 1, 0, &gid);
+	info->tx_attr->inject_size = init_attr.cap.max_inline_data;
+
 	ret = ibv_query_device(ctx, &device_attr);
-	if (ret)
-		return -errno;
+	if (ret) {
+		ret = -errno;
+		goto err3;
+	}
+
+	info->domain_attr->cq_cnt 		= device_attr.max_cq;
+	info->domain_attr->ep_cnt 		= device_attr.max_qp;
+	/* TODO find correct optimum values for ctx_cnt */
+	info->domain_attr->tx_ctx_cnt 		= device_attr.max_qp;
+	info->domain_attr->rx_ctx_cnt 		= device_attr.max_qp;
+	info->domain_attr->max_ep_tx_ctx 	= device_attr.max_qp;
+	info->domain_attr->max_ep_rx_ctx 	= device_attr.max_qp;
+	info->tx_attr->iov_limit 		= device_attr.max_sge;
+	info->tx_attr->rma_iov_limit 		= device_attr.max_sge;
+	info->rx_attr->iov_limit 		= device_attr.max_sge;
```

Comment:
> We should probably also set tx_attr->size and rx_attr->size based on
> device/queue pair attributes (device_attr.max_qp_wr?), rather than
> hardcoding a value.

This would result in the user *always* allocating the maximum size QP, which is highly undesirable.  The intent is to provide an 'optimal' value.  However, only the vendors can really provide that.  Selecting a hard-code value that attempts to bridge the gap between reasonable performance with minimal memory footprint seems ideal.  There's really no reason to construct QPs with queues 65,000 entries deep.